### PR TITLE
Feature/proxy-fmu

### LINKED ISSUE #628: Integrate with proxy-fmu
libcosim should integrate with proxy-server. The current fmu-proxy interface is similar, but needs to be adapted.

Q1: Should we move forward with this solution?

Q2: If yes, should proxy-server supply the network client as a library or should libcosim depend on Thrift and auto-generate the client sources as is currently done?

### PR DESCRIPTION
**Pull request by markaren**
Adds proxy-fmu support to libcosim. proxy-fmu is implemented in C++ and supports automatic spawning of new processes. Thus it replaces the current usage of the JVM based fmu-proxy solution.

Closes #628

### CODE REVIEW

**Comment by markaren** [comment]
Integration works locally at this point.

**Comment by markaren** [comment]
Not finding FMILibrary on linux when building proxyfmu is such a weird error. conan create works when triggered from proxyfmu repository. Why does it fail here? 
I tried running plain conan install without going through conan create first. Still the same error.

@kyllingstad any ideas?

**Comment by markaren** [comment]
Some progress.. However, now I get:

```
/usr/bin/ld: /home/runner/.conan/data/proxyfmu/0.2.0/osp/testing/package/92604b6200473d2d3e7f23059f7748858bfa20eb/lib/libproxyfmu-client.a(remote_slave.cpp.o): relocation R_X86_64_PC32 against symbol `_ZTVN8proxyfmu6thrift11BooleanReadE' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
```

**Comment by markaren** [comment]
OK, now it builds, but I had to add `PROPERTIES POSITION_INDEPENDENT_CODE ON` to the proxyfmu libraries (fmi and client).
Don't know the implications..

**Comment by kyllingstad** [comment]
Honestly, I don't know why they haven't made position-independent code the default in GCC yet. For most practical purposes, there are no negative implications, and PIC has the huge benefit that the code can later be linked into a into a dynamic library.

**Comment by markaren** [comment]
Might want to hold off reviewing until I got a proper test in place.

**Comment by markaren** [comment]
Mhh, as unittests are not run as part of the conan run, the proxyfmu test will never run in the CI.

**Comment by ljamt** [comment]
Looks like the [conan CI](https://github.com/open-simulation-platform/libcosim/actions/runs/774582370) fails.
Is it related to this[change](https://github.com/open-simulation-platform/proxy-fmu/commit/097fb289983dd61e945c70957554ea53420d1360) in proxy-fmu?

**Comment by markaren** [comment]
Doubt it. I ran with that update susessfully yesterday and here non proxy builds fails also.

**Comment by markaren** [comment]
>cheduled maintenance is currently in progress. We will provide updates as necessary.
Apr 22, 14:00 UTC - 18:00 UTC
Scheduled - Detail: The DNS record http://conan.bintray.com will be routed to the new ConanCenter repository.
Impact: Temporary instability, none lasting.

> If you receive this error, it is related to this maintenance:

**Comment by ljamt** [comment]
As a test I've build cosim-cli with this branch the conan package from this branch: 
`libcosim/0.8.0@osp/testing-feature_proxy-server`

Then reconfigured the [House demo configuration](https://github.com/open-simulation-platform/demo-cases/blob/master/house/OspSystemStructure.xml) to the following:
	
```
	<Simulators>
        <Simulator name="Clock" source="proxyfmu://localhost?file=Clock.fmu"/>
        <Simulator name="InnerWall" source="proxyfmu://localhost?file=InnerWall.fmu"/>
        <Simulator name="OuterWall1" source="proxyfmu://localhost?file=OuterWall1.fmu">
            <InitialValues>
                <InitialValue variable="T_outside">
                    <Real value="5.3"/>
                </InitialValue>
            </InitialValues>
        </Simulator>
        <Simulator name="OuterWall2" source="proxyfmu://localhost?file=OuterWall2.fmu">
            <InitialValues>
                <InitialValue variable="T_outside">
                    <Real value="4.9"/>
                </InitialValue>
            </InitialValues>
        </Simulator>
        <Simulator name="Room1" source="proxyfmu://localhost?file=Room1.fmu"/>
        <Simulator name="Room2" source="proxyfmu://localhost?file=Room2.fmu"/>
        <Simulator name="TempController" source="proxyfmu://localhost?file=TempController.fmu"/>
    </Simulators>
```

Testing with
`cosim run c:\dev\cse-demos\house\OspSystemStructure_.xml -d 10`

At first successful result:

```
@progress 1 1.000000 1.000000
@progress 2 2.000000 2.000000
@progress 3 3.000000 3.000000
@progress 4 4.000000 4.000000
@progress 5 5.000000 5.000000
@progress 6 6.000000 6.000000
@progress 7 7.000000 7.000000
@progress 8 8.000000 8.000000
@progress 9 9.000000 9.000000
@progress 10 10.000000 10.000000
[proxyfmu] Shutting down proxy for Clock::Clock done..
[proxyfmu] External proxy process for instance 'Clock' returned with status 0
[proxyfmu] Shutting down proxy for TempController::TempController done..
[proxyfmu] External proxy process for instance 'TempController' returned with status 0
[proxyfmu] Shutting down proxy for OuterWall1::OuterWall1 done..
[proxyfmu] External proxy process for instance 'OuterWall1' returned with status 0
[proxyfmu] Shutting down proxy for InnerWall::InnerWall done..
[proxyfmu] External proxy process for instance 'InnerWall' returned with status 0
[proxyfmu] Shutting down proxy for OuterWall2::OuterWall2 done..
[proxyfmu] External proxy process for instance 'OuterWall2' returned with status 0
[proxyfmu] Shutting down proxy for Room1::Room1 done..
[proxyfmu] External proxy process for instance 'Room1' returned with status 0
[proxyfmu] Shutting down proxy for Room2::Room2 done..
[proxyfmu] External proxy process for instance 'Room2' returned with status 0
```

But subsequent simulations gave:

```
Attempt 2:
Thrift: Thu May  6 16:42:20 2021 TServerSocket::listen() BIND 64466
Unhandled Exception reached the top of main: Could not bind: errno = 10013, application will now exit
[proxyfmu] External proxy process for instance 'Room1' returned with status 2
Thrift: Thu May  6 16:42:23 2021 TSocket::open() connect() <Host: localhost Port: 64466>: errno = 10061
Thrift: Thu May  6 16:42:25 2021 TSocket::open() connect() <Host: localhost Port: 64466>: errno = 10061

Attempt 3:
Thrift: Thu May  6 16:42:29 2021 TServerSocket::listen() BIND 53205
Unhandled Exception reached the top of main: Could not bind: errno = 10013, application will now exit
[proxyfmu] External proxy process for instance 'OuterWall1' returned with status 2
Thrift: Thu May  6 16:42:32 2021 TSocket::open() connect() <Host: localhost Port: 53205>: errno = 10061
Thrift: Thu May  6 16:42:34 2021 TSocket::open() connect() <Host: localhost Port: 53205>: errno = 10061

Attempt 4:
Thrift: Thu May  6 16:42:49 2021 TServerSocket::listen() BIND 53308
Unhandled Exception reached the top of main: Could not bind: errno = 10013, application will now exit
[proxyfmu] External proxy process for instance 'InnerWall' returned with status 2
Thrift: Thu May  6 16:42:52 2021 TSocket::open() connect() <Host: localhost Port: 53308>: errno = 10061
Thrift: Thu May  6 16:42:54 2021 TSocket::open() connect() <Host: localhost Port: 53308>: errno = 10061

```

Curious to hear if you see the same.

**Comment by markaren** [comment]
Hmm, runs fine here. You might have been unlucky with the port selection?

**Comment by ljamt** [comment]
> Hmm, runs fine here (same case). You might have been unlucky with the port selection?

You are probably right. For me, this specific case works ~3 out of 10 attempts.

File: src/cosim/proxy/remote_slave.cpp
Comment (kyllingstad) [inline src/cosim/proxy/remote_slave.cpp:195]: I haven't looked into the functions you're calling here, but I just want to note that one cannot throw exceptions from destructors(*). So if there is a non-neglectible chancethat these functions may throw, you may want to wrap the calls in a `try`-`catch`-statement and log a message on error.

(*) Or, you can, but it causes the program to immediately crash via `std::terminate()`.

File: tools/housekeeping
Comment (kyllingstad) [inline tools/housekeeping:16]: You should remove the line where `excludeFromClangFormat` is being used further down too.

File: tools/tmp_cleanup
Comment (kyllingstad) [inline tools/tmp_cleanup:15]: Is this still needed? (If so, maybe change "fmu-proxy" to "proxy-fmu" in the line above.)

**Review by kyllingstad — CHANGES_REQUESTED** [review]
This looks really good, great work! I have some comments about error handling and formatting that I'd like you to address, but otherwise, I think this is good to go!

(Sorry for taking so long before doing the review, btw.)

File: tools/tmp_cleanup
Comment (markaren) [reply tools/tmp_cleanup:15]: May be left behind by wrapped FMUs

**Review by kyllingstad — APPROVED** [review]
Great job!

**Comment by markaren** [comment]
Can merge this once a new release has been made from master and @ljamt has vetted it. I'd also like to have #637 merged first, so I can make use of it in this PR.

**Comment by ljamt** [comment]
> Can merge this once a new release has been made from master and @ljamt has vetted it. I'd also like to have #637 merged first, so I can make use of it in this PR.

Done testing and approving the PR now. Great work!!
Apologies for the delay.

**Review by ljamt — APPROVED** [review]

**Comment by markaren** [comment]
There is still the issue that this feature is not part of the unit testing though...

**Comment by ljamt** [comment]
> There is still the issue that this feature is not part of the unit testing though...

This PR includes two unit test files. What feature is not part of the unit testing?

**Comment by markaren** [comment]
Conan builds are not tested through the CI.

**Comment by markaren** [comment]
Enabled using testing for this feature. Tests passes, but some other issues appeared for gcc8.

**Pull request merged**
